nibabel/cmdline/diff.py

```diff
@@ -0,0 +1,121 @@
+#!python
+# emacs: -*- mode: python-mode; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the NiBabel package for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Quick summary of the differences among a set of neuroimaging files
+"""
+from __future__ import division, print_function, absolute_import
+
+import sys
+from optparse import OptionParser, Option
+
+import numpy as np
+
+import nibabel as nib
+import nibabel.cmdline.utils
+from nibabel.cmdline.utils import _err, verbose, table2string, ap, safe_get
+import fileinput
+
+__author__ = 'Yaroslav Halchenko & Christopher Cheng'
+__copyright__ = 'Copyright (c) 2017 NiBabel contributors'
+__license__ = 'MIT'
+
+
+def get_opt_parser():
+    # use module docstring for help output
+    p = OptionParser(
+        usage="%s [OPTIONS] [FILE ...]\n\n" % sys.argv[0] + __doc__,
+        version="%prog " + nib.__version__)
+
+    p.add_options([
+        Option("-v", "--verbose", action="count",
+               dest="verbose", default=0,
+               help="Make more noise.  Could be specified multiple times"),
+
+        Option("-H", "--header-fields",
+               dest="header_fields", default='all',
+               help="Header fields (comma separated) to be printed as well (if present)"),
+    ])
+
+    return p
+
+
+def diff_files(key, inputs):
+    diffs = []
+
+    for i in range(len(inputs)):
+        if i != len(inputs)-1:
+            temp_input_1 = inputs[i]
+            temp_input_2 = inputs[i+1]
+            if np.any(temp_input_1[key] != temp_input_2[key]):
+                if i != 0 and temp_input_2 != diffs:
+                    diffs.append(temp_input_1[key])
+                    diffs.append(temp_input_2[key])
+                else:
+                    diffs.append(temp_input_1[key])
+            elif type(temp_input_1[key]) != type(temp_input_2[key]):
+                if i != 0 and temp_input_2 != diffs:
+                    diffs.append(temp_input_1[key])
+                    diffs.append(temp_input_2[key])
+                else:
+                    diffs.append(temp_input_1[key])
+            else:
+                pass
+
+    # TODO: figure out a way to not have these erroneous outputs occur in the above loop
+    for a in range(len(diffs)-1):
+        for b in range(len(diffs)-1):
+            try:
+                if np.all(np.isnan(diffs[a])) and np.all(np.isnan(diffs[a+1])):
+                    del diffs[a]
+            except TypeError:
+                pass
+            if a and b != len(diffs)-1:
+                if np.any(diffs[a] == diffs[b]):
+                    del diffs[a]
+
+    if len(diffs) > 1:
+        return {key: diffs}
```

Comment:
I think above seems to be over-complicated a bit ;)
pretty much at the end you just want to know if any value (no need to encode key since also known outside) is different from any other value (e.g. the first one).  So, in human language 
```python
return any(different(x, values[0]) for x in values)
```
and you need just a simple "different" helper to reliably compare two values and outside of this function compose your "dict of diffs". I do not think you want to bother returning all pairwise differences, or some other attempt to be "smart" about the difference - would be even tricky to present, or later to associate with specific files in the output.  so [KISS](https://en.wikipedia.org/wiki/KISS_principle)